### PR TITLE
Add `from_pt` argument in `.from_pretrained`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR addresses https://github.com/huggingface/diffusers/issues/523 by attempting to add `from_pt` argument inside `.from_pretrained` function in `FlaxModelMixin` class. 
For now I followed the same logic as in `transformers` regarding this argument. However it seems that conversion tests are not implemented yet. Do we add these tests in this PR or it should be addressed in a future PR?

After this PR, as in `transformers` any model could be loaded in flax by doing
```
model = ...from_pretrained(..., from_pt=True)
```

cc @patrickvonplaten @mishig25 @patil-suraj @pcuenca 

I think we might need to add tests for this to make sure everything works as expected

### CODE REVIEW

**Comment by patil-suraj** [comment]
Awesome!

> I think we might need to add tests for this to make sure everything works as expected

#502

**Comment by patrickvonplaten** [comment]
That's a very good first draft! Thanks a lot :-) 

We will also have to rename all the weight names of https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_condition_flax.py to make it compatible as stated by @patil-suraj here: https://github.com/huggingface/diffusers/pull/502

@younesbelkada In short can you use the following code to iterate fast here? 

1) Create a dummy stable diffusion unet:

```python
from diffusers import UNet2DConditionModel, FlaxUNet2DConditionModel

model = UNet2DConditionModel(
    block_out_channels=(32, 64),
    layers_per_block=2,
    sample_size=32,
    in_channels=4,
    out_channels=4,
    down_block_types=("DownBlock2D", "CrossAttnDownBlock2D"),
    up_block_types=("CrossAttnUpBlock2D", "UpBlock2D"),
    cross_attention_dim=32,
)

with tempfile.TemporaryDirectory() as tmpdirname:
    model.save_pretrained(tmpdirname)
    flax_model = FlaxUNet2DConditionModel.from_pretrained(tmpdirname, from_pt=True
```

**Comment by patil-suraj** [comment]
Awesome!

> I think we might need to add tests for this to make sure everything works as expected

We should definitely add this for this!

Think we will need to wait a bit, we need to update the flax unet model to have the same weight names as in pt cf #502
once that's ready we can use that for the tests.

Also note that, we need handle the `nn.ModuleList` a bit differently here because the way flax treats module lists.

In torch for module list the keys are created like this `model.layers.0...`, `model.layers.1....`
In flax it's `model.layers_0....`, `model.layers_1...`.

In `tranformers` this is handled by creating a wrapper module (the `...LayerCollection` modules), to wrap the list inside the model to create the same naming structure for the keys. But that adds a lot of boilerplate so I'm not in favor of having this here.

We could instead use `regex` to rename the module list keys while converting to map them to correct names. For example:

```python
regex = r"\w+[.]\d+"

def rename_key(key):
   pats = re.findall(regex, key)
    for pat in pats:
        key = key.replace(pat, "_".join(pat.split(".")))
    return key
```

This will rename the `model.layers.0...` keys to flax format.

**Comment by patrickvonplaten** [comment]
@younesbelkada I think you can already get started on the PR here to make it work by renaming keys in https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_condition_flax.py and changing the conversion function as explained by Suraj and then we look into tests

**Comment by younesbelkada** [comment]
Perfect thanks for all the pointers! 
Yes I have just noticed that the Flax weights were named `layer_name_{index}` which is different from `transformers` - will rename them as suggested and test everting
Also as @mishig25 told me offline, there is no custom head logic so the function that I have copied will surely change!

**Comment by younesbelkada** [comment]
Few keys left before the full match! 
Just wondering how would you address the dropout issue? For now I just renamed the Sequential layer [here](https://github.com/huggingface/diffusers/blob/c1796efd5fc42c770dda3b800fdd9c411b634904/src/diffusers/models/attention_flax.py#L35) to `to_out_0` to match the PT weights

**Comment by patil-suraj** [comment]
> Just wondering how would you address the dropout issue? For now I just renamed the Sequential layer [here](https://github.com/huggingface/diffusers/blob/c1796efd5fc42c770dda3b800fdd9c411b634904/src/diffusers/models/attention_flax.py#L35) to to_out_0 to match the PT weights

Aah, didn't realise this. changing the name here makes sense.

Also now that we have JAX in diffusers, let's avoid using `dropout` or non module functions in `nn.Sequential`, better if we could just avoid it.

File: src/diffusers/models/unet_blocks_flax.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def setup(self):
         self.attentions = attentions
 
         if self.add_downsample:
-            self.downsample = FlaxDownsample2D(self.out_channels, dtype=self.dtype)
+            self.downsamplers_0 = FlaxDownsample2D(self.out_channels, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_blocks_flax.py:58]: This should be a list like in PyTorch

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 2]
```diff
@@ -214,6 +214,13 @@ def __call__(
             When returning a tuple, the first element is the sample tensor.
         """
         # 1. time
+        timesteps = timestep
```
Comment (patil-suraj) [inline src/diffusers/models/unet_2d_condition_flax.py:217]: why ?

**Comment by younesbelkada** [comment]
Here is a working v1 of the conversion script! The test now pass with `atol=rtol=4e-02` and is super flaky. Here is the script I use to check the correctness of the logits:
```
#!/usr/bin/env python3
from diffusers import UNet2DConditionModel, FlaxUNet2DConditionModel
import tempfile
import torch
import numpy as np
import jax.numpy as jnp

model = UNet2DConditionModel(
    block_out_channels=(32, 64),
    layers_per_block=2,
    sample_size=32,
    in_channels=4,
    out_channels=4,
    down_block_types=("DownBlock2D", "CrossAttnDownBlock2D"),
    up_block_types=("CrossAttnUpBlock2D", "UpBlock2D"),
    cross_attention_dim=32,
)

with tempfile.TemporaryDirectory() as tmpdirname:
    model.save_pretrained(tmpdirname)
    flax_model, flax_params = FlaxUNet2DConditionModel.from_pretrained(tmpdirname, from_pt=True)

sample = torch.rand(1, 4, model.config.sample_size, model.config.sample_size)
time = 1
text_emb = torch.rand(1, 77, model.config.sample_size)


# Step 1: Infer with the PT model
torch_output = model(sample, time, text_emb).sample


# Step 2: Infer with JAX model
flax_sample = jnp.array(sample.numpy())
flax_text_emb = jnp.array(text_emb.numpy())

flax_output = flax_model.apply({"params":flax_params}, jnp.transpose(flax_sample, (0, 2, 3, 1)), time, flax_text_emb).sample

# Step 3: Check that the values are close
converted_flax_output = torch.from_numpy(np.array(flax_output))
converted_flax_output = converted_flax_output.permute(0, 3, 1, 2)

torch.testing.assert_allclose(converted_flax_output, torch_output, rtol=4e-02, atol=4e-2)
```

File: src/diffusers/models/unet_2d_condition_flax.py (hunk 2 above)
Comment (younesbelkada) [reply src/diffusers/models/unet_2d_condition_flax.py:217]: This is because I copy-pasted the logic behind: https://github.com/huggingface/diffusers/blob/761f0297b08de18098e9cdbb82f36cb8911c3f8f/src/diffusers/models/unet_2d_condition.py#L216 , will change it!

File: src/diffusers/modeling_flax_pytorch_utils.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    pt_tuple_key: Tuple[str],
+    pt_tensor: np.ndarray,
+) -> (Tuple[str], np.ndarray):
+    """Rename PT weight names to corresponding Flax weight names and reshape tensor if necessary"""
+
+    # # conv norm or layer norm
+    # This is not really stable since any module that has the name 'scale'
+    # Will be affected. Maybe just check pt_tuple_key[-2] ?
+    renamed_pt_tuple_key = pt_tuple_key[:-1] + ("scale",)
+    if any("norm" in str_ for str_ in pt_tuple_key) and pt_tuple_key[-1] == "weight":
```
Comment (younesbelkada) [inline src/diffusers/modeling_flax_pytorch_utils.py:54]: This is quite hacky, we might need to change it

**Comment by patil-suraj** [comment]
Think the test is flaky because the unet is not in `eval` mode.

**Comment by younesbelkada** [comment]
Yes nice catch ! With eval mode the test pass with `4e-03` instead of `4e-02`

File: src/diffusers/models/unet_2d_condition_flax.py (hunk 2 above)
Comment (younesbelkada) [reply src/diffusers/models/unet_2d_condition_flax.py]: Proposed a fix in 7527ab185d744ac09948d86581bcae06e027f37c

File: src/diffusers/models/unet_blocks_flax.py (hunk 1 above)
Comment (younesbelkada) [reply src/diffusers/models/unet_blocks_flax.py:58]: I can see that there is always one element in the list in the PyTorch modeling file, is it expected that this will change in the future to insert more modules?

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 4]
```diff
@@ -214,6 +214,12 @@ def __call__(
             When returning a tuple, the first element is the sample tensor.
         """
         # 1. time
+        if not isinstance(timesteps, jnp.ndarray):
+            timesteps = jnp.array([timesteps], dtype=jnp.int32)
+        elif isinstance(timesteps, jnp.ndarray) and len(timesteps.shape) == 0:
+            timesteps = timesteps.astype(dtype=jnp.float32)
+            timesteps = timesteps[None]
```
Comment (patil-suraj) [inline src/diffusers/models/unet_2d_condition_flax.py:221]: maybe use `expand_dims` as it will be cleaner.

File: src/diffusers/models/unet_blocks_flax.py (hunk 1 above)
Comment (patil-suraj) [reply src/diffusers/models/unet_blocks_flax.py:58]: Good point!  @patrickvonplaten why did we use list for downsamplers ?

File: src/diffusers/modeling_flax_pytorch_utils.py (hunk 3 above)
Comment (patil-suraj) [reply src/diffusers/modeling_flax_pytorch_utils.py:54]: this is what I used in my repo, could we try adapting it like this ? Here's the full script https://github.com/patil-suraj/stable-diffusion-jax/blob/main/stable_diffusion_jax/convert_diffusers_to_jax.py

```python
if (
            "norm" in pt_key
            and (pt_tuple_key[-1] == "bias")
            and (pt_tuple_key[:-1] + ("bias",) not in random_flax_state_dict)
            and (pt_tuple_key[:-1] + ("scale",) in random_flax_state_dict)
        ):
            pt_tuple_key = pt_tuple_key[:-1] + ("scale",)
        elif pt_tuple_key[-1] in ["weight", "gamma"] and pt_tuple_key[:-1] + ("scale",) in random_flax_state_dict:
            pt_tuple_key = pt_tuple_key[:-1] + ("scale",)
```

File: src/diffusers/models/unet_blocks_flax.py (hunk 1 above)
Comment (younesbelkada) [reply src/diffusers/models/unet_blocks_flax.py:58]: The only solution I see is to Wrap it in a `LayerCollection` module but this goes against what we want to achieve [here](https://github.com/huggingface/diffusers/pull/527#issuecomment-1249313185) - also I agree that this solution adds a lot of boilerplate

**Review by patil-suraj — COMMENTED** [review]
Very cool, thanks for working on this! 

For tests, we need to add the common test mixin for flax and then we can add tests for this. This could be done in a follow-up PR.

File: src/diffusers/models/unet_2d_condition_flax.py (hunk 4 above)
Comment (younesbelkada) [reply src/diffusers/models/unet_2d_condition_flax.py]: Fixed in 1facd9fa4b9fb164be779c9cb5feeb60c23c4895 !

File: src/diffusers/modeling_flax_pytorch_utils.py (hunk 3 above)
Comment (younesbelkada) [reply src/diffusers/modeling_flax_pytorch_utils.py]: Thanks a lot for the snippet! Adapted the changes in 5973e435410222f4ad702226d04ee800fe5ee13d
Now it also supports converting embedding layers (which was not the case before your comment)

**Comment by younesbelkada** [comment]
Thanks a lot for the comments @patil-suraj ! Should have addressed them now 💪 
I am still worried that the tests still pass only with `atol=rtol=4e-03`, as far as I can see from the Pytorch script there is no stochasticity involved in the forward pass

**Comment by patrickvonplaten** [comment]
Very nice! Will take a closer look tomorrow. 

Could you try to apply the same changes to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/vae_flax.py tomorrow? :-)

**Comment by younesbelkada** [comment]
@patrickvonplaten @patil-suraj 
I just tried with VAE and can confirm the test pass with lower tolerance than Unet (`2e-05`)
Here is the script I use to compare
```
#!/usr/bin/env python3
from diffusers import AutoencoderKL, FlaxAutoencoderKL
import tempfile
import torch
import numpy as np
import jax.numpy as jnp

model = AutoencoderKL(
    in_channels = 3,
    out_channels = 3,
    sample_size=32,
    latent_channels= 4,
    norm_num_groups= 32,
)

model = model.eval()

with tempfile.TemporaryDirectory() as tmpdirname:
    model.save_pretrained(tmpdirname)
    flax_model, flax_params = FlaxAutoencoderKL.from_pretrained(tmpdirname, from_pt=True)

sample = torch.rand(1, 3, model.config.sample_size, model.config.sample_size)
time = 1


# Step 1: Infer with the PT model
torch_output = model(sample).sample


# Step 2: Infer with JAX model
flax_sample = jnp.array(sample.numpy())

flax_output = flax_model.apply({"params":flax_params}, flax_sample).sample

# Step 3: Check that the values are close
converted_flax_output = torch.from_numpy(np.array(flax_output))

torch.testing.assert_allclose(converted_flax_output, torch_output, rtol=2e-05, atol=2e-05)
```
One small detail though, I can see that the shapes are slightly inconsistent between the VAE and Unet, the first model expects a shape of the format `n_channels, img_w, img_h` whereas the Unet expects an input of the format `img_w, img_h, n_channels`. It seems that the VAE does the permutation operations internally. Do you think we have to change either one of the model expected input shape to make it consistent across models? Happy to do that just let me know ;)

**Comment by younesbelkada** [comment]
Added a commit 4cad1aeb4aeb224402dad13c018a5d42e96267f6 for shapes consistency that we could revert if the current input/output shapes are expected, the test now pass with this script (similar as Unet):

```
#!/usr/bin/env python3
from diffusers import AutoencoderKL, FlaxAutoencoderKL
import tempfile
import torch
import numpy as np
import jax.numpy as jnp

model = AutoencoderKL(
    in_channels = 3,
    out_channels = 3,
    sample_size=32,
    latent_channels= 4,
    norm_num_groups= 32,
)

model = model.eval()

with tempfile.TemporaryDirectory() as tmpdirname:
    model.save_pretrained(tmpdirname)
    flax_model, flax_params = FlaxAutoencoderKL.from_pretrained(tmpdirname, from_pt=True)

sample = torch.rand(1, 3, model.config.sample_size, model.config.sample_size)
time = 1


# Step 1: Infer with the PT model
torch_output = model(sample).sample


# Step 2: Infer with JAX model
flax_sample = jnp.array(sample.numpy())

flax_output = flax_model.apply({"params":flax_params}, jnp.transpose(flax_sample, (0, 2, 3, 1))).sample

# Step 3: Check that the values are close
converted_flax_output = torch.from_numpy(np.array(flax_output))
converted_flax_output = converted_flax_output.permute(0, 3, 1, 2)

# # Step 3: Check that the values are close
# converted_flax_output = torch.from_numpy(np.array(flax_output))

torch.testing.assert_allclose(converted_flax_output, torch_output, rtol=2e-05, atol=2e-05)
```

**Comment by younesbelkada** [comment]
@patil-suraj as discussed offline I have just fixed the channels issue!
Now the test passes with:
```
#!/usr/bin/env python3
from diffusers import UNet2DConditionModel, FlaxUNet2DConditionModel
import tempfile
import torch
import numpy as np
import jax.numpy as jnp

model = UNet2DConditionModel(
    block_out_channels=(32, 64),
    layers_per_block=2,
    sample_size=32,
    in_channels=4,
    out_channels=4,
    down_block_types=("DownBlock2D", "CrossAttnDownBlock2D"),
    up_block_types=("CrossAttnUpBlock2D", "UpBlock2D"),
    cross_attention_dim=32,
)

model = model.eval()

with tempfile.TemporaryDirectory() as tmpdirname:
    model.save_pretrained(tmpdirname)
    flax_model, flax_params = FlaxUNet2DConditionModel.from_pretrained(tmpdirname, from_pt=True)

sample = torch.rand(1, 4, model.config.sample_size, model.config.sample_size)
time = 1
text_emb = torch.rand(1, 77, model.config.sample_size)


# Step 1: Infer with the PT model
torch_output = model(sample, time, text_emb).sample


# Step 2: Infer with JAX model
flax_sample = jnp.array(sample.numpy())
flax_text_emb = jnp.array(text_emb.numpy())

flax_output = flax_model.apply({"params":flax_params}, flax_sample, time, flax_text_emb).sample

# Step 3: Check that the values are close
converted_flax_output = torch.from_numpy(np.array(flax_output))

torch.testing.assert_allclose(converted_flax_output, torch_output, rtol=4e-03, atol=4e-03)
```

Channels first!

**Review by patil-suraj — APPROVED** [review]
Thanks a lot! 

Works perfectly for unet model. For `FlaxAutoencoderKL` we need to rename some keys, will open a follow-up a PR for that. Merging!
